source/lac/trilinos_vector.cc

```diff
@@ -244,7 +237,19 @@ namespace TrilinosWrappers
       // with the map in v, and generate the vector.
       if (allow_different_maps == false)
         {
-          if (vector->Map().SameAs(v.vector->Map()) == false)
+          // check equality for MPI communicators: We can only choose the fast
+          // version in case the underlying Epetra_MpiComm object is the same,
+          // otherwise we might access an MPI_Comm object that has been
+          // deleted
+#ifdef DEAL_II_WITH_MPI
+          const Epetra_MpiComm *my_comm = dynamic_cast<const Epetra_MpiComm *>(&vector->Comm());
+          const Epetra_MpiComm *v_comm = dynamic_cast<const Epetra_MpiComm *>(&v.vector->Comm());
+          const bool same_communicators = my_comm != NULL && v_comm != NULL &&
+                                          my_comm->DataPtr() == v_comm->DataPtr();
```

Comment:
Yes, it could happen. But I don't think we should care about the case where we reinit from a vector using a stale communicator, i.e., I think we should not consider the case where the vector we reinit from is defective.